p2p-sync/src/main/java/org/smartregister/p2p/presenter/P2PReceiverPresenter.java

```diff
@@ -165,65 +172,256 @@ public void onAuthorizationDetailsProvided(@NonNull Map<String, Object> authoriz
+                            }
+                        }, new GenericAsyncTask.OnFinishedCallback<Integer>() {
+                            @Override
+                            public void onSuccess(@Nullable Integer result) {
+                                if (result != null) {
+                                    Timber.e(view.getString(R.string.log_records_deleted), (int) result);
+                                }
+
+                                sendLastReceivedRecords(new ArrayList<P2pReceivedHistory>());
+                            }
+
+                            @Override
+                            public void onError(Exception e) {
+                                Timber.e(view.getString(R.string.log_error_occurred_trying_to_delete_p2p_received_history_on_device)
+                                        , sendingDevice.getDeviceId());
+                                disconnectAndReset(endpointId);
+                            }
+                        });
+
                     }
                 } else {
-                    onConnectionAuthorizationRejected("Authorization details sent by receiver are invalid");
+                    // This is a new device we should save it
+                    Tasker.run(new Callable<Void>() {
+                        @Override
+                        public Void call() throws Exception {
+                            registerSendingDevice(basicDeviceDetails);
+                            return null;
+                        }
+                    }, new GenericAsyncTask.OnFinishedCallback<Void>() {
+                        @Override
+                        public void onSuccess(@Nullable Void result) {
+                            sendLastReceivedRecords(new ArrayList<P2pReceivedHistory>());
+                        }
+
+                        @Override
+                        public void onError(Exception e) {
+                            Timber.e(e);
+                            view.showToast(view.getString(R.string.an_error_occurred_trying_to_save_new_sender_details), Toast.LENGTH_LONG);
+
+                            disconnectAndReset(endpointId);
+                        }
+                    });
                 }
             }
-        }
+
+            @Override
+            public void onError(Exception e) {
+                Timber.e(e);
+                disconnectAndReset(endpointId);
+            }
+        });
+    }
+
+    @Override
+    public void onPayloadTransferUpdate(@NonNull String endpointId, @NonNull PayloadTransferUpdate update) {
+        // Do nothing for now
     }
 
     @Override
     public void onDisconnected(@NonNull String endpointId) {
-        Timber.e(view.getString(R.string.log_endpoint_lost), endpointId);
-        resetState();
-        startAdvertisingMode();
+        if (getCurrentPeerDevice() != null && endpointId.equals(getCurrentPeerDevice().getEndpointId())) {
```

Comment:
Consider adding else to do cleanup

Reply:
The else is not supposed to be possible, if it is, then due to the memory leak whose issue is here https://github.com/OpenSRP/android-p2p-sync/issues/39. 

The app was crashing without the if since the code was being run even after exiting the activity and the ConnectionsClient is holding on to the callbacks passed

 If the [google issue](https://issuetracker.google.com/issues/129429944) has no reply by the time I am done with the remaining issues and no other solution in site, then we'll have to use WeakReferences for the view (Or clean it up) but the API might still hold onto the listener until it makes the last call

The else happening, means that the callback was called although we are not connected to the endpoint or have already disconnected